### PR TITLE
Create Functional Cache for Custom State SSZ

### LINKED ISSUE #4138: Improve hash_tree_root for process_slot
We have seen inefficiency when we `process_slot` due to `hash_tree_root(state)` is slow but the only item changed in the state is `slot`. It wouldn't be that slow with proper caching branches of the state tree structure. This issue tracks how we can improve. We've seen multiple incidents where RPC context times out due to slowness of `process_slot`

### PR DESCRIPTION
**Pull request by rauljordan**
Follow-up to #4077 this resolves #4138 

---

# Description

**Write why you are making the changes in this pull request**

Currently, our SSZ does not maintain a proper merkleization for arrays of roots, making it difficult to hash the beacon state repeatedly even if a single element changed. 

**Write a summary of the changes you are making**

This PR adds a caching layer using ristretto which properly supports merkleizing the beacon state's biggest fields. This approach leads to a massive improvement at runtime, nearly eliminating all the flames in a pprof flame graph regarding hash tree root.

### CODE REVIEW

File: shared/stateutil/arrays.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+var (
+	fastSumHashKey = bytesutil.ToBytes32([]byte("hash_fast_sum64_key"))
+	leavesCache    = make(map[string][][]byte)
+	layersCache    = make(map[string][][][]byte)
+	lock           sync.Mutex
+)
+
+func arraysRoot(roots [][]byte, fieldName string) ([32]byte, error) {
+	lock.Lock()
+	if _, ok := layersCache[fieldName]; !ok {
```
Comment (0xKiwi) [inline shared/stateutil/arrays.go]: Does this need to be it's own lock sequence? Would it be possible to group it with the other lock below?

File: shared/stateutil/attestations.go [hunk 2]
```diff
@@ -4,13 +4,44 @@ import (
...
+
+	copy(enc[16:48], data.BeaconBlockRoot)
+
+	// Source epoch and root.
+	sourceEpochBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(sourceEpochBuf, data.Source.Epoch)
+	copy(enc[48:56], sourceEpochBuf)
+	copy(enc[56:88], data.Source.Root)
+
+	// Target..
```
Comment (0xKiwi) [inline shared/stateutil/attestations.go]: Extra period here

File: shared/stateutil/arrays.go (hunk 1 above)
Comment (prestonvanloon) [reply shared/stateutil/arrays.go]: I think it's OK to have minimal locking groups.

File: shared/stateutil/arrays.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		layersCache[fieldName] = make([][][]byte, depth+1)
+	}
+	lock.Unlock()
+
+	hashKeyElements := make([]byte, len(roots)*32)
+	leaves := make([][]byte, len(roots))
+	emptyKey := highwayhash.Sum(hashKeyElements, fastSumHashKey[:])
+	bytesProcessed := 0
+	changedIndices := make([]int, 0)
+	lock.Lock()
```
Comment (prestonvanloon) [inline shared/stateutil/arrays.go]: This can be a RLock

File: shared/stateutil/arrays.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		for i := 0; i < len(changedIndices); i++ {
+			rt, err = recomputeRoot(changedIndices[i], chunks, fieldName)
+			if err != nil {
+				return [32]byte{}, err
+			}
+		}
+		return rt, nil
+	}
+
+	hashKey := highwayhash.Sum(hashKeyElements, fastSumHashKey[:])
```
Comment (prestonvanloon) [inline shared/stateutil/arrays.go]: Is highway hash for sha256 faster than github.com/minio/sha256-simd? 

I think we should use only one and use `hashutil.Hash`

File: shared/stateutil/arrays.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	"sync"
+
+	"github.com/minio/highwayhash"
+	"github.com/protolambda/zssz/merkle"
+	"github.com/prysmaticlabs/prysm/shared/bytesutil"
+	"github.com/prysmaticlabs/prysm/shared/hashutil"
+)
+
+var (
+	fastSumHashKey = bytesutil.ToBytes32([]byte("hash_fast_sum64_key"))
```
Comment (prestonvanloon) [inline shared/stateutil/arrays.go]: This is the same key has sum64, would it cause collisions? Can we use hashutil to keep everything together?

**Pull request merged**